### ISSUE TITLE
On NodeJS When loosing connection, the socket takes long time to close

### ISSUE THREAD

**Issue by genadis** [stomp-js/stompjs#226]
I'm using `ws` to pollyfil the Websocket on Node.
On timeout of heartbeatIncoming https://github.com/stomp-js/stompjs/blob/master/src/stomp-handler.ts#L262 [_closeWebsocket](https://github.com/stomp-js/stompjs/blob/1121fd901669f9290f95766f65d5e3eca06c591d/src/stomp-handler.ts#L268) is called.

I've enabled debug to see the timing and I can see that there are multiple `did not receive server activity for the last ${delta}ms` before the socket is actually closed.

`_closeWebsocket` uses `_webSocket.close()` which is not immediate, it waits for timeout of the socket and can take 30 seconds.

According to `ws` docs [how-to-detect-and-close-broken-connections](https://github.com/websockets/ws#how-to-detect-and-close-broken-connections) they recommend to use `terminate()` method to immediately close connection.

It's ok to wait a long timeout when we voluntarily close connection, but if the connection is disconnected due to failure, and the application has no way of knowing about it for a long period of time, messages are lost and the ping/pong keep alive mechanism looses it's purpose.

To reproduce:
1. successfully connect 
2. disconnect the network interface of the client
3. Even thought the `heartbeatIncoming` expires, the `connected` flag is still `true` and the application have no way of knowing that the connection is lost.
4. the close is detected after more than 30 seconds and only then the `connected` flag updated to false.

Simple solution could be adding to `stomp-handler`:
```
public _terminateWebsocket() {
    this._webSocket.onmessage = () => { }; // ignore messages
    if (this._webSocket.terminate) {
        this._webSocket.terminate();
    } else {
        this._webSocket.close();
     }
 }
```
and to use it on keep alive timeout instead of regular close.

Following https://github.com/stomp-js/stompjs/pull/224 any websocket wrapper would be able to implement it's own optional `terminate` method.

**Comment by genadis** [comment]
@kum-deepak I'll be happy to update the https://github.com/stomp-js/stompjs/pull/224 PR with the necessary changes.

**Comment by kum-deepak** [comment]
This one is actually little deeper in my opinion. I faced this issue a while back in the browser. I could not totally make up my mind that what would be correct behavior.

When connection closed explicitly by using `deactivate` the current behavior seems correct.

However, if the connection is getting closed because of HeartBeat issues, it become tricky. During that period, if a message arrives, should we deliver it for processing or drop. Eventually I decided to drop those (`this._webSocket.onmessage = () => { };` is a recent addition to the code).

There is an associated race condition, if someone called `decativate` immediately followed by `activate` there may be two WebScokets active at the same time - rare but possible.

The solution that you are suggesting will work for NodeJS with `ws`, we need to define ideal behavior and solve the larger issue.

For this tactical solution, I will prefer exposing a callback without adding this code into the library itself.

**Comment by kum-deepak** [comment]
Let us work on a new PR :+1:

**Comment by genadis** [comment]
I believe the most important thing is to expose to the user of the library the correct state. We already have the flag `connected` so if it would be to represent the most accurate state, the user of the library can take care of the rest for example:
* queue messages to publish, and send to broker on reconnect.
* design app tolerant for lost messages on subscribe, or alternatively use manual ack and broker that supports persistence and delivery retries.

The dangerous state is when the app "thinks" it's `connected` to broker while in fact it is not.

So to address the topics you brought up:

> When connection closed explicitly by using deactivate the current behavior seems correct.

I believe so as well. the `connected` will be false immediately. So it's fine if the close callbacks takes time. And the user of the library can manage state if it's important for the app for next activate.

> There is an associated race condition, if someone called decativate immediately followed by activate there may be two WebScokets active at the same time - rare but possible.

That's true, but the user of the app is aware of that, and if that condition is problematic the app should manage the state and not call activate until close is finished.
We could change the api of activate and deactivate to Promise based or add callbacks to make it more explicit for the user of the library.

> However, if the connection is getting closed because of HeartBeat issues, it become tricky. During that period, if a message arrives, should we deliver it for processing or drop. Eventually I decided to drop those (this._webSocket.onmessage = () => { }; is a recent addition to the code).

I believe there is an issue for how it is currently handled. the same code is used to handle heartbeat issues as for deactivate which is:
```
public _closeWebsocket() {
    this._webSocket.onmessage = () => { }; // ignore messages
    this._webSocket.close();
}
```
The problem is for sending and for receiving messages. For sending it's more prominent though. The problem is because the `connected` flag does not reflect the state. As I've mentioned `this._webSocket.close();` is not immediate it can take a long time till the onclose event is fired and the `connected` flag changes to false.
during all this time:
* For receiving: messages are dropped and the app "thinks" it's still connected. It is less prominent than sending because for critical apps that use manual acks and broker that supports redelivery the messages won't be lost indefinitely.
* For sending: The app thinks it's `connected` and will continue to publish messages that would be just lost. If the `connected` would to reflect the true state it could have queued them instead.

For the above reasons I've suggested to implement separate method for this condition of heartbeat issues and following your notes this could be (added `this._cleanUp();`): 
```typescript
public _terminateWebsocket() {
    this._webSocket.onmessage = () => { }; // ignore messages
    this._cleanUp(); // immediately indicate that the connection is lost 
    if (this._webSocket.terminate) {
        this._webSocket.terminate();
    } else {
        this._webSocket.close();
     }
 }
```
And on the websocket interface:
```typescript
export interface IStompSocket {
  /**
   * Optional
   * Terminates the connection forcefully.
   */
  terminate?(): void;
  // ....
}
```

And maybe `forceDisconnect` of the client should use `this._stompHandler._terminateWebsocket();` as well.


> For this tactical solution, I will prefer exposing a callback without adding this code into the library itself.

The above suggestion is not tactical only for Node, it will help with predictability for the browser as well.

**Comment by kum-deepak** [comment]
This all makes sense. I think we can start work on this in a PR.

Thinking ahead:

- For `deactivate` it should resolve immediately if it is not connected. If connected, it should resolve after it is disconnected and cleaned up.
- For `activate`, I do not have clear idea, if we need it to be aync. If yes then when it should resolve. There may not be a single successful connection. 
- For `activate`, one option can be that it resolves immediately under normal circumstances. If it is still deactivating, it waits till the deactivation is complete and it complete activation phase (which is just setup, does not imply that it got connected).

In any case let us start the work, we can target to release as 6.x.

**Comment by kum-deepak** [comment]
Released as part of 6.0.0.

**Comment by davidkhala** [comment]
@kum-deepak Sorry for my jump into this thread while searching for async activate option, I support that for `activate` should resolve immediately under normal circumstances, just like what we did in async _connect already. And now the way we call async _connect without await as a step in `activate`, is so a bad practice.
